lightning-invoice/src/utils.rs

```diff
@@ -437,6 +439,207 @@ mod test {
+			&nodes[0],
+			HashSet::new(),
+		);
+	}
+
+	#[test]
+	fn test_channels_with_lower_inbound_capacity_than_invoice_amt_hints_filtering() {
+		let chanmon_cfgs = create_chanmon_cfgs(3);
+		let node_cfgs = create_node_cfgs(3, &chanmon_cfgs);
+		let node_chanmgrs = create_node_chanmgrs(3, &node_cfgs, &[None, None, None]);
+		let nodes = create_network(3, &node_cfgs, &node_chanmgrs);
+		let chan_1_0 = create_unannounced_chan_between_nodes_with_value(&nodes, 1, 0, 100_000, 0, InitFeatures::known(), InitFeatures::known());
+		let chan_2_0 = create_unannounced_chan_between_nodes_with_value(&nodes, 2, 0, 1_000_000, 0, InitFeatures::known(), InitFeatures::known());
+
+		// As the invoice amt is 1 msat above chan_1_0's inbound capacity, it shouldn't be included
+		let mut scid_aliases_99_000_001_msat = HashSet::new();
+		scid_aliases_99_000_001_msat.insert(chan_2_0.0.short_channel_id_alias.unwrap());
+
+		match_invoice_routes(
+			Some(99_000_001),
+			&nodes[0],
+			scid_aliases_99_000_001_msat,
+		);
+
+		// As the invoice amt is exactly at chan_1_0's inbound capacity, it should be included
+		let mut scid_aliases_99_000_000_msat = HashSet::new();
+		scid_aliases_99_000_000_msat.insert(chan_1_0.0.short_channel_id_alias.unwrap());
+		scid_aliases_99_000_000_msat.insert(chan_2_0.0.short_channel_id_alias.unwrap());
+
+		match_invoice_routes(
+			Some(99_000_000),
+			&nodes[0],
+			scid_aliases_99_000_000_msat,
+		);
+
+		// As the invoice amt is above all channels' inbound capacity, they will still be included
+		let mut scid_aliases_2_000_000_000_msat = HashSet::new();
+		scid_aliases_2_000_000_000_msat.insert(chan_1_0.0.short_channel_id_alias.unwrap());
+		scid_aliases_2_000_000_000_msat.insert(chan_2_0.0.short_channel_id_alias.unwrap());
+
+		match_invoice_routes(
+			Some(2_000_000_000),
+			&nodes[0],
+			scid_aliases_2_000_000_000_msat,
+		);
+
+		// An invoice with no specified amount should include all channels in the route hints.
+		let mut scid_aliases_no_specified_amount = HashSet::new();
+		scid_aliases_no_specified_amount.insert(chan_1_0.0.short_channel_id_alias.unwrap());
+		scid_aliases_no_specified_amount.insert(chan_2_0.0.short_channel_id_alias.unwrap());
+
+		match_invoice_routes(
+			None,
+			&nodes[0],
+			scid_aliases_no_specified_amount,
+		);
+	}
+
+	fn match_invoice_routes<'a, 'b: 'a, 'c: 'b>(
+		invoice_amt: Option<u64>,
+		invoice_node: &Node<'a, 'b, 'c>,
+		mut chan_ids_to_match: HashSet<u64>
+	){
+		let invoice = create_invoice_from_channelmanager_and_duration_since_epoch(
+			&invoice_node.node, invoice_node.keys_manager, Currency::BitcoinTestnet, invoice_amt, "test".to_string(),
+			Duration::from_secs(1234567)).unwrap();
+		let hints = invoice.private_routes();
+
+		assert_eq!(hints.len(), chan_ids_to_match.len());
+
+		for hint in hints {
+			let hint_short_chan_id = (hint.0).0[0].short_channel_id;
+			assert!(chan_ids_to_match.contains(&hint_short_chan_id));
+			chan_ids_to_match.remove(&hint_short_chan_id);
```

Comment:
Did you want to check that `chan_ids_to_match` is empty after iterating through the hints? This could be more useful than the earlier length check because you could include in the assertion failure which channels had been missed.

Reply:
True!